egui_glium/src/painter.rs

```diff
@@ -1,192 +1,386 @@
+        ) {
+            unsafe {
+                gl.delete_texture(old_tex);
+            }
+        }
         self.egui_texture_version = Some(texture.version);
     }
 
+    unsafe fn prepare_painting(
+        &mut self,
+        gl_window: &glutin::WindowedContext<glutin::PossiblyCurrent>,
+        gl: &glow::Context,
+        pixels_per_point: f32,
+    ) -> (u32, u32) {
+        gl.enable(glow::SCISSOR_TEST);
+        // egui outputs mesh in both winding orders:
+        gl.disable(glow::CULL_FACE);
+
+        gl.enable(glow::BLEND);
+        gl.blend_equation(glow::FUNC_ADD);
+        gl.blend_func_separate(
+            // egui outputs colors with premultiplied alpha:
+            glow::ONE,
+            glow::ONE_MINUS_SRC_ALPHA,
+            // Less important, but this is technically the correct alpha blend function
+            // when you want to make use of the framebuffer alpha (for screenshots, compositing, etc).
+            glow::ONE_MINUS_DST_ALPHA,
+            glow::ONE,
+        );
+
+        let glutin::dpi::PhysicalSize {
+            width: width_in_pixels,
+            height: height_in_pixels,
+        } = gl_window.window().inner_size();
+        let width_in_points = width_in_pixels as f32 / pixels_per_point;
+        let height_in_points = height_in_pixels as f32 / pixels_per_point;
+
+        gl.viewport(0, 0, width_in_pixels as i32, height_in_pixels as i32);
+
+        gl.use_program(Some(self.program));
+
+        // The texture coordinates for text are so that both nearest and linear should work with the egui font texture.
+        // For user textures linear sampling is more likely to be the right choice.
+        gl.uniform_2_f32(Some(&self.u_screen_size), width_in_points, height_in_points);
+        gl.uniform_1_i32(Some(&self.u_sampler), 0);
+        gl.active_texture(glow::TEXTURE0);
+
+        gl.bind_vertex_array(Some(self.va));
+        gl.bind_buffer(glow::ARRAY_BUFFER, Some(self.vb));
+        gl.bind_buffer(glow::ELEMENT_ARRAY_BUFFER, Some(self.eb));
+        (width_in_pixels, height_in_pixels)
+    }
+
     /// Main entry-point for painting a frame.
     /// You should call `target.clear_color(..)` before
     /// and `target.finish()` after this.
-    pub fn paint_meshes<T: glium::Surface>(
+    ///
+    /// The following OpenGL features will be set:
+    /// - Scissor test will be enabled
+    /// - Cull face will be disabled
+    /// - Blend will be enabled
+    ///
+    /// The scissor area and blend parameters will be changed.
+    ///
+    /// As well as this, the following objects will be rebound:
+    /// - Vertex Array
+    /// - Vertex Buffer
+    /// - Element Buffer
+    /// - Texture (and active texture will be set to 0)
+    /// - Program
+    ///
+    /// Please be mindful of these effects when integrating into your program, and also be mindful
+    /// of the effects your program might have on this code. Look at the source if in doubt.
```

Comment:
Great docs!

Reply:
I have a lot more documentation to write, so I'll keep this unresolved for now